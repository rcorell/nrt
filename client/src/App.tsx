import 'bootstrap/dist/css/bootstrap.min.css';
import React from 'react';
import { Provider } from 'react-redux';

import { Routes } from './components/routes/Routes';
import { Store } from './redux/store';

import { AppHeader, GlobalStyle } from './styles';

export class App extends React.Component {
  public render() {
      return (
          <>
              <GlobalStyle />
              <Provider store={Store}>
                  <AppHeader>Top 5 Daily</AppHeader>
                  <Routes />
              </Provider>
          </>
      );
  }
}
