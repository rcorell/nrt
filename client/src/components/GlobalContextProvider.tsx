import React, { useState } from 'react';

import { isAuthenticated } from 'src/api/api';

type GlobalContextType = {
    authenticated: boolean;
    setAuthenticated: (authenticated: boolean) => void;
};

export const GlobalContext = React.createContext<GlobalContextType>({
    authenticated: false,
    setAuthenticated: (_authenticated) => {}
});

export const GlobalContextProvider: React.FC = ({ children }) => {
    const [authenticated, setAuthenticated] = useState(isAuthenticated());

    const contextFunctions = { authenticated, setAuthenticated };

    return <GlobalContext.Provider value={contextFunctions}>{children}</GlobalContext.Provider>;
};
