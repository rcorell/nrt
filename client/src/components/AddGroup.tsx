import { useMutation } from '@apollo/client';
import React, { useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { createGroupMutationString } from 'src/api/api';
import { CreateGroupMutation, CreateGroupMutationVariables } from 'src/api/__generated__/CreateGroupMutation';
import { setBrowserTitle } from 'src/utils';
import { AppError, AppForm, FormContainer } from 'src/styles/form';

export const AddGroup: React.FC = () => {
    const [description, setDescription] = useState('');
    const [name, setName] = useState('');

    setBrowserTitle('Add Group');

    const [createGroup, { error }] = useMutation<CreateGroupMutation, CreateGroupMutationVariables>(
        createGroupMutationString,
        {
            variables: { description, name },
            onCompleted: ({ createGroup }) => {
                createGroup.id += 0;
            }
        }
    );

    const isFormInvalid = () => {
        return name.length < 1;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        createGroup();
    };

    const errorDisplay = () => {
        if (error) {
            return <div>Error: {JSON.stringify(error)}</div>;
        }

        return null;
    };

    return (
        <FormContainer>
            <h1>Add Group</h1>
            {errorDisplay()}
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label>Name</Form.Label>
                    <Form.Control
                        autoFocus
                        id="name"
                        onChange={(e) => setName(e.target.value)}
                        size="lg"
                        value={name}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label>Description</Form.Label>
                    <Form.Control
                        id="description"
                        onChange={(e) => setDescription(e.target.value)}
                        size="lg"
                        value={description}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Add Group
                </Button>
            </AppForm>
            <AppError></AppError>
        </FormContainer>
    );
};
