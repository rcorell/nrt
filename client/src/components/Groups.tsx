import React from 'react';
import * as _ from 'lodash';
import { useMutation, useQuery } from '@apollo/client';
import MaterialTable from 'material-table';

import { fetchGroupsQueryString, fetchUserQueryString, joinGroupMutationString } from 'src/api/api';
import { FetchGroupsQuery } from 'src/api/__generated__/FetchGroupsQuery';
import { FetchUserQuery } from 'src/api/__generated__/FetchUserQuery';
import { JoinGroupMutation, JoinGroupMutationVariables } from 'src/api/__generated__/JoinGroupMutation';
import { setBrowserTitle } from 'src/utils';

// const renderJoinColumn = (rowData: any) => {
//     if (userGroups.contains(rowData.name)) {
//         return <p>Joined</p>;
//     }

//     return <p>Join</p>;
// };

const COLUMN_DEFINITIONS = [
    { title: 'Name', field: 'name' },
    { title: 'Description', field: 'description' },
    {
        field: 'join',
        title: ''
    }
];

export const Groups: React.FC = () => {
    setBrowserTitle('Groups');

    const userQuery = useQuery<FetchUserQuery>(fetchUserQueryString);
    const allGroups = useQuery<FetchGroupsQuery>(fetchGroupsQueryString);

    const [joinGroup] = useMutation<JoinGroupMutation, JoinGroupMutationVariables>(joinGroupMutationString, {
        onCompleted: () => {
            // console.log('woo');
            userQuery.refetch();
        },
        onError: () => {
            // https://github.com/apollographql/apollo-client/issues/7167
        }
    });

    if (allGroups.loading || userQuery.loading) {
        return <div>Loading...</div>;
    }

    if (allGroups.error || userQuery.error) {
        return <div>Error: {JSON.stringify(allGroups.error) + JSON.stringify(userQuery.error)}</div>;
    }

    const userGroupIds = userQuery!.data!.user.groups.map((group) => {
        return group!.id;
    });

    const groups = _.cloneDeep(allGroups.data!.groups).map((group) => {
        return {
            ...group,
            join: userGroupIds.includes(group.id) ? (
                'Joined'
            ) : (
                <button onClick={() => joinGroup({ variables: { groupId: Number(group.id) } })}>Join!</button>
            )
        };
    });

    return <MaterialTable title="Groups" columns={COLUMN_DEFINITIONS} data={groups} />;
};
