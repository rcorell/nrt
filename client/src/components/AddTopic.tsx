import { useMutation, useQuery } from '@apollo/client';
import React, { useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { CreateTopicMutation, CreateTopicMutationVariables } from 'src/api/__generated__/CreateTopicMutation';
import { FetchUserQuery } from 'src/api/__generated__/FetchUserQuery';
import { createTopicMutationString, fetchUserQueryString } from 'src/api/api';
import { LOADING_TEXT } from 'src/components/shared';
import { AppError, AppForm, FormContainer } from 'src/styles/form';
import { setBrowserTitle } from 'src/utils';

export const AddTopic: React.FC = () => {
    const [description, setDescription] = useState('');
    const [groupId, setGroupId] = useState('');
    const [title, setTitle] = useState('');

    const userQuery = useQuery<FetchUserQuery>(fetchUserQueryString);

    const [createTopic, { error }] = useMutation<CreateTopicMutation, CreateTopicMutationVariables>(
        createTopicMutationString,
        {
            variables: { groupId, title, description },
            onError: () => {
                // RTL bug
            }
        }
    );

    const isFormInvalid = () => {
        return title.length < 1;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        createTopic();
    };

    const errorStatus = () => {
        if (error) {
            return <div>Error: ${JSON.stringify(error)}</div>;
        }

        return null;
    };

    setBrowserTitle('Add Topic');

    if (userQuery.loading || !userQuery.data) {
        return <div>{LOADING_TEXT}</div>;
    }

    const groupOptions = userQuery.data.user.groups.map((group) => (
        <option key={group!.id} value={group!.id}>
            {group!.name}
        </option>
    ));

    return (
        <FormContainer>
            <h1>Add Topic</h1>
            {errorStatus()}
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label id="groupLabel">Group</Form.Label>
                    <Form.Control
                        as="select"
                        aria-labelledby="groupLabel"
                        id="group"
                        onChange={(e) => setGroupId(e.target.value)}
                        size="lg"
                    >
                        {groupOptions}
                    </Form.Control>
                </Form.Group>

                <Form.Group>
                    <Form.Label id="topicLabel">Topic</Form.Label>
                    <Form.Control
                        autoFocus
                        aria-labelledby="topicLabel"
                        id="topic"
                        onChange={(e) => setTitle(e.target.value)}
                        size="lg"
                        value={title}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label id="descriptionLabel">Description</Form.Label>
                    <Form.Control
                        id="description"
                        aria-labelledby="descriptionLabel"
                        onChange={(e) => setDescription(e.target.value)}
                        size="lg"
                        value={description}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Submit Topic
                </Button>
            </AppForm>
            <AppError></AppError>
        </FormContainer>
    );
};
