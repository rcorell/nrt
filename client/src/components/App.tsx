import 'bootstrap/dist/css/bootstrap.min.css';

import { ApolloProvider } from '@apollo/client';
import React from 'react';

import { GlobalContextProvider } from 'src/components/GlobalContextProvider';
import { apolloClient } from 'src/api/api';
import { Routes } from 'src/components/Routes';
import { AppHeader, GlobalStyle } from 'src/styles';

export const App: React.FC = () => {
    return (
        <>
            <GlobalStyle />
            <GlobalContextProvider>
                <ApolloProvider client={apolloClient}>
                    <AppHeader>Top 5 Daily</AppHeader>
                    <Routes />
                </ApolloProvider>
            </GlobalContextProvider>
        </>
    );
};
