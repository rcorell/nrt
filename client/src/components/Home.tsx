import { useQuery } from '@apollo/client';
import React from 'react';

import {
    FetchUserWithGroupTopicsQuery,
    FetchUserWithGroupTopicsQuery_user_groups
} from 'src/api/__generated__/FetchUserWithGroupTopicsQuery';
import { fetchUserWithGroupTopicsQuery } from 'src/api/api';
import { LOADING_TEXT } from 'src/components/shared';
import { HomePage } from 'src/styles/home';
import { setBrowserTitle } from 'src/utils';

export const Home: React.FC = () => {
    const query = useQuery<FetchUserWithGroupTopicsQuery>(fetchUserWithGroupTopicsQuery);

    setBrowserTitle('Home');

    if (query.loading) {
        return <b>{LOADING_TEXT}</b>;
    }
    if (query.error) {
        return <div>{JSON.stringify(query.error)}</div>;
    }

    const groupTopics = (group: FetchUserWithGroupTopicsQuery_user_groups | null) => {
        return (
            <div key={group!.id}>
                <h2>{group!.name}</h2>
                <ul>
                    {group!.topics.map((topic) => (
                        <li key={topic!.id}>{topic!.title}</li>
                    ))}
                </ul>
            </div>
        );
    };

    return <HomePage>{query!.data!.user.groups.map((group) => groupTopics(group))}</HomePage>;
};
