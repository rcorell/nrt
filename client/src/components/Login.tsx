import { useMutation } from '@apollo/client';
import { navigate } from 'hookrouter';
import React, { useContext, useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { LoginMutation, LoginMutationVariables } from 'src/api/__generated__/LoginMutation';
import { loginMutation } from 'src/api/api';
import { GlobalContext } from 'src/components/GlobalContextProvider';
import { AppError, AppForm, FormContainer } from 'src/styles/form';
import { setBrowserTitle } from 'src/utils';

export const Login: React.FC = () => {
    const { setAuthenticated } = useContext(GlobalContext);

    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');

    const [login, { error }] = useMutation<LoginMutation, LoginMutationVariables>(loginMutation, {
        onCompleted: (result) => {
            if (result?.login) {
                localStorage.setItem('token', result.login.token!);
                setAuthenticated(true);
                navigate('/');
            }
        },
        onError: () => {
            // https://github.com/apollographql/apollo-client/issues/7167
        },
        variables: { email, password }
    });

    setBrowserTitle('Login');

    const isFormInvalid = () => {
        return email.length < 7 || password.length === 0;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        login();
    };

    const authStatus = () => {
        if (error) {
            return <AppError>Error: {error.message}</AppError>;
        }

        return null;
    };

    return (
        <FormContainer>
            <h1>Login</h1>
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label id="emailLabel">Email</Form.Label>
                    <Form.Control
                        aria-labelledby="emailLabel"
                        autoFocus
                        id="email"
                        onChange={(e) => setEmail(e.target.value)}
                        size="lg"
                        type="email"
                        value={email}
                        placeholder={'email'}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label id="passwordLabel">Password</Form.Label>
                    <Form.Control
                        aria-labelledby="passwordLabel"
                        id="password"
                        onChange={(e) => setPassword(e.target.value)}
                        size="lg"
                        type="password"
                        value={password}
                        placeholder={'password'}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Log In
                </Button>
            </AppForm>
            <AppError>{authStatus()}</AppError>
        </FormContainer>
    );
};
