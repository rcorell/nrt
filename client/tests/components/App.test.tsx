import { App } from 'src/components/App';

describe('App', () => {
    it('matches snapshot', () => {
        global.expectComponentToMatchSnapshot(App);
    });
});
