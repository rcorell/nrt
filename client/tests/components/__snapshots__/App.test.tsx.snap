// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<div>
  <div
    class="sc-eCImvq eFmfDu"
  >
    Top 5 Daily
  </div>
  <div
    class="sc-iCfLBT kRgVyw"
  >
    <div
      class="sc-furvIG"
    >
      <a
        href="/login"
      >
        Log in
      </a>
       | 
      <a
        href="/signup"
      >
        Sign up
      </a>
    </div>
    <div
      class="sc-pVTma"
    />
  </div>
  <div
    class="sc-gKckTs ldzOLX"
  >
    <b>
      Loading...
    </b>
  </div>
</div>
`;
